### ISSUE TITLE
Support rooting at solution tsconfig files

### ISSUE THREAD

**Issue by Gelio** [neovim/nvim-lspconfig#940]
- `nvim --version`: 

<details>
<summary>
NVIM v0.5.0-dev+1339-g5d9c56012
</summary>
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/work/neovim/neovim/build/config -I/home/runner/work/neovim/neovim/src -I/home/runner/work/neovim/neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim/neovim/build/src/nvim/auto -I/home/runner/work/neovim/neovim/build/include
Compiled by runner@fv-az121-644

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/home/runner/work/neovim/neovim/build/nvim.AppDir/usr/share/nvim"

Run :checkhealth for more info
</details>

- nvim-lsp version(commit hash): 6a25eb0d3fc77cd1b9ff11446dc52d50bc482155
- Is the problem isolated to a particular language server: yes
- What language server (If the problem is related to a specific language server): [tsserver](https://github.com/neovim/nvim-lspconfig/blob/master/CONFIG.md#tsserver)
- Can you reproduce this behavior on other language server clients (vscode, languageclient-neovim, coc.nvim, etc.): It works as expected in VSCode. I suppose this is a problem with setting the correct `root`, thus, a config problem
- Can you reproduce this behavior on other language servers offered in the nvim-lspconfig repo? (pyls -> pyright):
- Operating system/version: Ubuntu 20.10

### How to reproduce the problem from neovim startup

1. Have tsserver installed and configured in your neovim
2. Clone https://github.com/RyanCavanaugh/learn-a
3. Set up the `learn-a` repo (`npm install && npx lerna bootstrap && npx tsc -b packages`)
4. `nvim packages/pkg1/src/foo.ts`
5. Get the references for `fn`

### Actual behaviour

There is only a single reference for `fn`:
![image](https://user-images.githubusercontent.com/889383/119683284-d0cb4200-be43-11eb-84fd-662e6ab3c0ff.png)

This is most likely because the started language server is rooted in `packages/pkg1`:
![image](https://user-images.githubusercontent.com/889383/119683428-ee001080-be43-11eb-8801-af0e59a94295.png)

(there is a `tsconfig.json` and `package.json` there, so that's expected given the default `root_dir`:
https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/tsserver.lua#L14-L17

I tried to come up with an alternative `root_dir` that would honor [solution tsconfig files](https://www.typescriptlang.org/docs/handbook/project-references.html#overall-structure), but I did not manage to find any easy way :worried: Perhaps setting `root_dir` should be customizable for those projects in which we want to use a solution file, so I could exactly specify which `root_dir` I want tsserver to start with?

### Expected behaviour

There are 3 references for `fn`, since it is used 3 times across the whole solution. It does work fine in VSCode:
![image](https://user-images.githubusercontent.com/889383/119683670-20117280-be44-11eb-8663-60b966613be3.png)

There should be a single language server for `tsserver` rooted at `packages`, since that's the place that has [the solution's `tsconfig.json`](https://github.com/RyanCavanaugh/learn-a/blob/master/packages/tsconfig.json)

### Minimal init.vim or init.lua and code sample

There is nothing unordinary about the minimal `init.lua`, just the default `tsserver` config:

<details>
-- install packer
local fn = vim.fn

local install_path = '/tmp/nvim/site/pack/packer/start/packer.nvim'

vim.cmd [[set packpath=/tmp/nvim/site]]

local function load_plugins()
  local use = require('packer').use
  require("packer").startup(
    {
      function()
        use 'wbthomason/packer.nvim'
        use "neovim/nvim-lspconfig"
      end,
      config = {package_root = '/tmp/nvim/site/pack'}
    }
  )
end

_G.load_config = function()
  vim.lsp.set_log_level("trace")
  local nvim_lsp = require('lspconfig')
  local on_attach = function(_, bufnr)
    local function buf_set_keymap(...) vim.api.nvim_buf_set_keymap(bufnr, ...) end
    local function buf_set_option(...) vim.api.nvim_buf_set_option(bufnr, ...) end

    buf_set_option('omnifunc', 'v:lua.vim.lsp.omnifunc')

    -- Mappings.
    local opts = { noremap=true, silent=true }
    buf_set_keymap('n', 'gD', '<Cmd>lua vim.lsp.buf.declaration()<CR>', opts)
    buf_set_keymap('n', 'gd', '<Cmd>lua vim.lsp.buf.definition()<CR>', opts)
    buf_set_keymap('n', 'K', '<Cmd>lua vim.lsp.buf.hover()<CR>', opts)
    buf_set_keymap('n', 'gi', '<cmd>lua vim.lsp.buf.implementation()<CR>', opts)
    buf_set_keymap('n', '<C-k>', '<cmd>lua vim.lsp.buf.signature_help()<CR>', opts)
    buf_set_keymap('n', '<space>wa', '<cmd>lua vim.lsp.buf.add_workspace_folder()<CR>', opts)
    buf_set_keymap('n', '<space>wr', '<cmd>lua vim.lsp.buf.remove_workspace_folder()<CR>', opts)
    buf_set_keymap('n', '<space>wl', '<cmd>lua print(vim.inspect(vim.lsp.buf.list_workspace_folders()))<CR>', opts)
    buf_set_keymap('n', '<space>D', '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)
    buf_set_keymap('n', '<space>rn', '<cmd>lua vim.lsp.buf.rename()<CR>', opts)
    buf_set_keymap('n', 'gr', '<cmd>lua vim.lsp.buf.references()<CR>', opts)
    buf_set_keymap('n', '<space>e', '<cmd>lua vim.lsp.diagnostic.show_line_diagnostics()<CR>', opts)
    buf_set_keymap('n', '[d', '<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>', opts)
    buf_set_keymap('n', ']d', '<cmd>lua vim.lsp.diagnostic.goto_next()<CR>', opts)
    buf_set_keymap('n', '<space>q', '<cmd>lua vim.lsp.diagnostic.set_loclist()<CR>', opts)
  end

  -- Add the server that troubles you here
  local name = "tsserver"
  if not name then
    print("You have not defined a server name, please edit minimal_init.lua")
  end
  if not nvim_lsp[name].document_config.default_config.cmd and (not cmd) then
    print([[You have not defined a server default cmd for a server
      that requires it please edit minimal_init.lua]])
  end

  nvim_lsp[name].setup {
      cmd = cmd,
      on_attach = on_attach,
    }

  print([[You can find your log at $HOME/.cache/nvim/lsp.log. Please paste in a github issue under a details tag as described in the issue template.]])
  end

if fn.isdirectory(install_path) == 0 then
  fn.system({'git', 'clone', 'https://github.com/wbthomason/packer.nvim', install_path})
  load_plugins()
  require('packer').sync()
  vim.cmd 'autocmd User PackerComplete ++once lua load_config()'
else
  load_plugins()
  _G.load_config()
end
</details>

### Health check
<details>
<summary>Checkhealth result</summary>
health#lspconfig#check
========================================================================
## Checking language server protocol configuration
  - INFO: gopls: configuration checked.
  - INFO: rust_analyzer: configuration checked.
  - INFO: jsonls: configuration checked.
  - INFO: bashls: configuration checked.
  - INFO: cssls: configuration checked.
  - INFO: stylelint_lsp: configuration checked.
  - INFO: tsserver: configuration checked.
  - INFO: svelte: configuration checked.
  - INFO: efm: configuration checked.
  - INFO: graphql: configuration checked.
</details>

### LSP log

 I don't think it's relevant, as the problem is with the `root_dir`, most likely

**Comment by lithammer** [comment]
If I understand the problem correctly, it sounds related to https://github.com/neovim/nvim-lspconfig/pull/652. But just so that I understand the problem correctly, you want `packages/` to be the `root_dir` rather than e.g. `packages/pkg1/` even if the latter also contains root markers (i.e. `tsconfig.conf` and `package.json`), correct?

**Comment by Gelio** [comment]
Exactly, you described it well :slightly_smiling_face: Both contain root markers, but it's the `packages/tsconfig.json` that should "win" in this case, since that is the config that includes all sub-packages

**Comment by Gelio** [comment]
Although, I don't see this as being related to https://github.com/neovim/nvim-lspconfig/pull/652. If I understand correctly, https://github.com/neovim/nvim-lspconfig/pull/652 solves a problem that is already solved for tsserver by calling `root_pattern` sequentially:
https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/tsserver.lua#L15-L16

The existing `root_dir` does not handle a situation in which the first found `tsconfig.json` is merely a config for the single project and there is some ancestor that contains a solution-wide `tsconfig.json` (which I think may be hard to detect without analyzing the contents of the files). This is why I suggested to have a manual override for the `root_dir` in such complex cases :thinking:

**Comment by mjlbach** [comment]
@Gelio you can configure project specific root directories, see the wiki entry https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings

**Comment by Gelio** [comment]
@mjlbach Thanks for the suggestion! I missed this page initially. I was able to configure my own project, as well as https://github.com/RyanCavanaugh/learn-a that I chose as a base for reproduction steps.

![image](https://user-images.githubusercontent.com/889383/119701736-fd3b8a00-be54-11eb-8fee-df2eea107852.png)


One thing I found is that the example `.nvimrc` in https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings is missing a return:
```diff
-  nvim_lsp.rust_analyzer.setup { root_dir= function() vim.fn.getcwd() end; }
+  nvim_lsp.rust_analyzer.setup { root_dir= function() return vim.fn.getcwd() end; }
```

or at least it did not work for me without the `return`.

Personally, being a lua beginner, I had some trouble figuring out that I also need to forward the `on_attach` function from my `init.lua`. I assume that calling `setup` overwrites the previously set `on_attach`.

I could try to update that wiki page the added `return` and add more detailed instructions how to forward the `on_attach` from `init.lua`. Let me know if you're open for PRs, or you'd rather handle one/both fixes yourself

---

This is a good enough workaround for the issue. Maybe there is a more general solution to the problem that does not require creating custom `.nvimrc`s. For me, the issue can be closed, but I'll let you decide if you want to try to have a more intelligent fix

Thanks for the help!
